### PR TITLE
run commands on comments

### LINKED ISSUE #1: Make app work as a one time executable
Make the application work as a one time execution.

By this, it means that it is not a server. It simply turns on, gets a command, runs such command and then exits.

It should work with some kind of hardcoded environment variables (for this first run)

### LINKED ISSUE #2: Make app get information from a comment
Make the app capable of reading a GitHub comment (as the event triggering it) and fetch the required data from it.

### PR DESCRIPTION
**Pull request by Bullrich**
Added ability to parse comments when it's triggered by a `issue_comment` event.

Updated documentation showing how to execute the action's code.

Added composite actions to trigger the commands when passed to a matrix and push changes to the branch.

Resolves #1 and #2

### CODE REVIEW

File: .github/workflows/cmd-action.yml [hunk 1]
```diff
@@ -1,15 +1,45 @@
...
           GITHUB_TOKEN: ${{ github.token }}
+      - name: Export branch name
+        run: echo "branch=$(git rev-parse --abbrev-ref HEAD)" >> "$GITHUB_OUTPUT"
+        id: branch
+
+  cmd-run:
+    needs: [cmd-check]
+    if: ${{ github.event.issue.pull_request }}
+    continue-on-error: true
+    # We set the current machine here (to differ between ours and generic ones)
```
Comment (rzadp) [inline .github/workflows/cmd-action.yml]: I don't get this comment, could you elaborate/rephrase?

File: run/action.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        env:
+          COMMAND: ${{ inputs.command }}
+      - name: Report failure
+        if: ${{ failure() }}
+        shell: bash
+        run: echo "Command failed!"
+      - name: Check changes
+        shell: bash
+        run: |
+          git config --global --add safe.directory /__w/${{ github.event.repository.name }}/${{ github.event.repository.name }}
```
Comment (rzadp) [inline run/action.yml]: Would it be cleaner to use `.` or `$(pwd)`?

File: README.md [hunk 3]
```diff
@@ -10,27 +10,52 @@ Create a file named `.github/workflows/cmd-action.yml` and add the following:
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        command: ${{ fromJson(needs.cmd-check.outputs.commands) }}
+    name: Run command
+    steps:
+      - uses: paritytech/cmd-action/run@main
+        with:
+          branch: ${{ needs.cmd-check.outputs.branch }}
+          command: ${{ matrix.command }}
```
Comment (mutantcornholio) [inline README.md:43]: This sounds a bit extensive for the repository-side action. Sounds like something we will have to update frequently.  
Can we have this code inside of this repo, and just set the parameters?

File: .github/workflows/cmd-action.yml [hunk 4]
```diff
@@ -1,15 +1,45 @@
...
+        if: ${{ github.event.issue.pull_request }}
+        env:
+          GITHUB_TOKEN: ${{ github.token }}
       - uses: paritytech/cmd-action@main
+        id: cmd
         with: 
           commands-directory: '.github/scripts'
           GITHUB_TOKEN: ${{ github.token }}
+      - name: Export branch name
+        run: echo "branch=$(git rev-parse --abbrev-ref HEAD)" >> "$GITHUB_OUTPUT"
```
Comment (mutantcornholio) [inline .github/workflows/cmd-action.yml]: Why do we need to parse branch here? Don't we have it in the variables?

File: .github/workflows/cmd-action.yml (hunk 1 above)
Comment (Bullrich) [reply .github/workflows/cmd-action.yml]: It‘s from the original proposal, so we can define custom self hosted machines for some more demanding jobs.

I will eventually add some more things here, but I would say to ignore it for now

File: run/action.yml (hunk 2 above)
Comment (Bullrich) [reply run/action.yml]: That‘s a good idea, I‘ll try that!

File: .github/workflows/cmd-action.yml (hunk 4 above)
Comment (Bullrich) [reply .github/workflows/cmd-action.yml]: No, when it‘s generated from an issue comment we only have the PR number, but not the branch name, so I need to obtain it somehow. I could also do `gh pr view —json headRefName —jq „.headRefName“` to obtain the same result.

File: README.md (hunk 3 above)
Comment (Bullrich) [reply README.md:43]: I‘m afraid I‘m not following you. Could you please rephrase it?

Comment (mutantcornholio) [reply README.md:43]: I meant the whole action. There's logic in here, with checking out the PR branch, doing some calculations. I'd expect something more simple to be put in the repositories that use the action. 

We are already reusing `paritytech/cmd-action/run@main`, is it possible to move some of this logic in this repo, and not spread it outside?

File: run/action.yml (hunk 2 above)
Comment (Bullrich) [reply run/action.yml]: I tried but end up reverting it as the running directory is `/home/runner/work/polkadot-sdk/polkadot-sdk`

Comment (rzadp) [reply run/action.yml]: Do you have a log of a failed run because of this? 🤔

Comment (Bullrich) [reply run/action.yml]: Here is tge test run: https://github.com/paritytech-stg/polkadot-sdk/actions/runs/8927469091/job/24520953847

**Comment by rzadp** [comment]
And do you have a failed run log when you dont configure this safe directory at all?On 2 May 2024, at 19:33, Javier Bullrich ***@***.***> wrote:﻿
@Bullrich commented on this pull request.



In run/action.yml:
> +        # Here it would get the script from previous step
+        run: |
+          echo "$COMMAND" >> command.sh
+          chmod +x command.sh
+          ./command.sh
+          rm command.sh
+        env:
+          COMMAND: ${{ inputs.command }}
+      - name: Report failure
+        if: ${{ failure() }}
+        shell: bash
+        run: echo "Command failed!"
+      - name: Check changes
+        shell: bash
+        run: |
+          git config --global --add safe.directory /__w/${{ github.event.repository.name }}/${{ github.event.repository.name }}

Here is tge test run: https://github.com/paritytech-stg/polkadot-sdk/actions/runs/8927469091/job/24520953847

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because your review was requested.Message ID: ***@***.***>

This communication and any attachments are confidential.

File: run/action.yml (hunk 2 above)
Comment (rzadp) [reply run/action.yml]: And do you have a failed run log when you dont configure the safe directory at all?

Comment (Bullrich) [reply run/action.yml]: As I was having many issues, I decided to replace it with [stefanzweifel/git-auto-commit-action](https://github.com/stefanzweifel/git-auto-commit-action).

File: README.md (hunk 3 above)
Comment (Bullrich) [reply README.md:43]: I now simplified it into it's own composite action.

Except the git push as that can be dependent on the docker image and I believe it's better to let the author customize it.

File: run/action.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  using: "composite"
+  steps:
+      - name: Download repo
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.branch }}
+      - name: Generate script
+        shell: bash
+        # Here it would get the script from previous step
+        run: |
```
Comment (rzadp) [inline run/action.yml]: I believe something like this would work in one line without the need for the file:

```
bash -c "$COMMAND"
```

**Review by rzadp — APPROVED** [review]

File: run/action.yml (hunk 5 above)
Comment (Bullrich) [reply run/action.yml]: I worked flawlessly! Great recommendation!

**Pull request merged**